ts/private/ts_project_worker.js

```diff
@@ -374,82 +465,46 @@ function createProgram(args, initialInputs) {
-        // invalidate failed lookups cache of tsc.
-        //
-        // Assume that `filePath` is `bazel-out/darwin_arm64-fastbuild/bin/feature1/index.d.ts`
-        
-        // First, in case the directory containing `index.d.ts is` in failed lookups, we need to 
-        // get the directory watcher for `bazel-out/darwin_arm64-fastbuild/bin/feature1`
-        // which is `bazel-out/darwin_arm64-fastbuild/bin`
-        // and invoke it with `bazel-out/darwin_arm64-fastbuild/bin/feature1`
-        const dir = path.dirname(filePath);
-        const directoryWatcherForGrandParentDirectory = getDirectoryWatcherForPath(dir)
-        directoryWatcherForGrandParentDirectory?.(dir);
-
-        // Then, get the directory watcher for `bazel-out/darwin_arm64-fastbuild/bin/feature1/index.d.ts`
-        // which is `bazel-out/darwin_arm64-fastbuild/bin/feature1`
-        // and invoke it with `bazel-out/darwin_arm64-fastbuild/bin/feature1/index.d.ts`
-        const directoryForParentDirectory = getDirectoryWatcherForPath(filePath)
-        directoryForParentDirectory?.(filePath);
-
-        // Then finally a fileWatcher for the filePath and invoke it.
-        const fileWatcherCallback = fileWatchers.get(filePath);
-        fileWatcherCallback?.(kind);
     }
 
     function watchDirectory(directoryPath, callback, recursive, options) {
-        directoryPath = path.relative(execRoot, directoryPath);
         // since rules_js runs everything under bazel-out we shouldn't care about anything outside of it.
-        if (!directoryPath.startsWith('bazel-out')) {
+        if (!directoryPath.startsWith(execRoot)) {
             return { close: noop };
         }
-        debuglog(`watchDirectory ${directoryPath}`);
+        const close = filesystemTree.watchDirectory(
+            path.relative(execRoot, directoryPath),
+            (input) => callback(path.join(execRoot, input))
+        );
 
-        const cb = (input) => callback(path.join(execRoot, input))
-
-        const callbacks = directoryWatchers.get(directoryPath) || new Set();
-        callbacks.add(cb)
-        directoryWatchers.set(directoryPath, callbacks);
-
-        return {
-            close: () => {
-                const callbacks = directoryWatchers.get(directoryPath);
-                callbacks.delete(cb);
-                if (callbacks.size === 0) {
-                    directoryWatchers.delete(directoryPath)
-                }
-            },
-        };
+        return {close};
     }
 
     function watchFile(filePath, callback, interval) {
-        debuglog(`watchFile ${filePath} ${interval}`);
-        const relativeFilePath = path.relative(execRoot, filePath);
-        fileWatchers.set(relativeFilePath, (kind) => callback(filePath, kind));
-        return { close: () => fileWatchers.delete(relativeFilePath) };
+        if (!path.isAbsolute(filePath)) {
+            filePath = path.resolve(filePath);
+        } else if (!filePath.startsWith(execRoot)) {
```

Comment:
If the path was relative we never need to do this? Should it be `if` instead of `else if`?

Reply:
The path is sometimes relative. The only instance I came across was file watcher opened against tsconfig.json. the reason why https://github.com/aspect-build/rules_ts/runs/7643046681?check_suite_focus=true#step:7:72 failed at https://github.com/aspect-build/rules_ts/pull/108/commits/83b946e3ff2016ebf88bc228732b7c8e407dce0e

Reply:
and `path.relative(filePath)` won't ever start with `execRoot`?

Reply:
ah, you are right. if typescript ever passed something like "../../../../" it might be out of execRoot when normalized. will fix it!

Reply:
fixed.